examples/research_projects/codeparrot/scripts/minhash_deduplication.py

```diff
@@ -0,0 +1,207 @@
+        di.add(filename, min_hash)
+
+    # Returns a List[Cluster] where Cluster is List[str] with the filenames.
+    return di.get_duplicate_clusters()
+
+
+def jaccard_similarity(code1: str, code2: str) -> float:
+    """Compute the Jaccard similarity of two code snippets."""
+    tokens1 = get_tokens(code1)
+    tokens2 = get_tokens(code2)
+    return len(tokens1 & tokens2) / len(tokens1 | tokens2)
+
+
+_shared_dataset = None
+
+
+def _find_cluster_extremes_shared(cluster):
+    """
+    Find a reduced cluster such that each code in the origin cluster is similar to at least one code in the reduced cluster.
+    """
+    extremes = []
+    for element1 in cluster:
+        code1 = _shared_dataset[element1["base_index"]]["content"]
+        for element2 in extremes:
+            code2 = _shared_dataset[element2["base_index"]]["content"]
+            if jaccard_similarity(code1, code2) >= 0.85:
+                element2["copies"] += 1
+                break
+        else:
+            element1["copies"] = 1
+            extremes.append(element1)
+    return extremes
+
+
+def multipro_find_extremes(cluster_list, dataset):
+    global _shared_dataset
+    _shared_dataset = dataset
+    extremes_list = []
+    with mp.Pool() as pool:
+        for extremes in tqdm(
+            pool.imap_unordered(
+                _find_cluster_extremes_shared,
+                cluster_list,
+            ),
+            total=len(cluster_list),
+        ):
+            extremes_list.append(extremes)
+    return extremes_list
+
+
+def deduplicate_dataset(dataset: Type[Dataset]) -> Tuple[Type[Dataset], List[List[Dict]]]:
+    """Deduplicate the dataset using minhash and jaccard similarity.
+    This function first generate duplicate clusters, then each cluster
+    is reduced to the extremes that are similar to the other elements in the cluster.
+    Codes are called similar if their Jaccard similarity is greater than 0.85.
+
+    Parameters
+    ----------
+    dataset : Type[Dataset]
+        The dataset to deduplicate.
+
+    Returns
+    -------
+    ds_dedup : Type[Dataset]
+        The deduplicated dataset.
+    duplicate_clusters : List[List[Dict]]
+        The list of duplicate clusters.
+        Each cluster is a list of dicts with the following keys:
+        - base_index : int
+            The index of the code in the original dataset.
+        - repo_name : str
+        - path : str
+        - copies : int
+            The number of copies of the code in the cluster. (find_cluster_extremes)
+        - is_extreme : bool
+            Whether the code is an extreme in the cluster.
+        All the codes in the cluster are removed from the dataset except the extremes.
+    """
+    duplicate_clusters = make_duplicate_clusters(dataset)
+    duplicate_indices = set(x["base_index"] for cluster in duplicate_clusters for x in cluster)
+    # todo: make this multiprocessing, pay attention to memory usage
```

Comment:
Is this `TODO` done with the new function below?

Reply:
exactly !